File: assets/scripts/core/ToolCard.ts
export interface ToolCard {
  puzzleId: string;
  stage: number;
  front: {
    toolName: string;
    scene: string;
    wisdomCrystal: string;
  };
  back: {
    coreAction: string;
    whenToUse: string[];
    realLifeExamples: string[];
    commonTraps: string;
  };
  unlockedAt: number;
}

export type ToolCardDraft = Omit<ToolCard, "unlockedAt">;

export type ToolCardValidationResult =
  | { ok: true; value: ToolCard }
  | { ok: false; errors: string[] };

export function createToolCard(draft: ToolCardDraft, unlockedAt = Date.now()): ToolCard {
  return {
    puzzleId: draft.puzzleId,
    stage: draft.stage,
    front: {
      toolName: draft.front.toolName,
      scene: draft.front.scene,
      wisdomCrystal: draft.front.wisdomCrystal
    },
    back: {
      coreAction: draft.back.coreAction,
      whenToUse: [...draft.back.whenToUse],
      realLifeExamples: [...draft.back.realLifeExamples],
      commonTraps: draft.back.commonTraps
    },
    unlockedAt
  };
}

export function validateToolCard(value: unknown): ToolCardValidationResult {
  const errors: string[] = [];

  if (!isRecord(value)) {
    return { ok: false, errors: ["tool card must be an object"] };
  }

  requireNonEmptyString(value, "puzzleId", errors);
  if (!Number.isInteger(value.stage) || (value.stage as number) < 1 || (value.stage as number) > 5) {
    errors.push("stage must be an integer from 1 to 5");
  }

  validateFront(value.front, errors);
  validateBack(value.back, errors);

  if (typeof value.unlockedAt !== "number" || !Number.isFinite(value.unlockedAt)) {
    errors.push("unlockedAt must be a finite number");
  }

  if (errors.length > 0) {
    return { ok: false, errors };
  }

  return { ok: true, value: value as unknown as ToolCard };
}

function validateFront(value: unknown, errors: string[]): void {
  if (!isRecord(value)) {
    errors.push("front must be an object");
    return;
  }

  requireNonEmptyString(value, "toolName", errors, "front.toolName");
  requireNonEmptyString(value, "scene", errors, "front.scene");
  requireNonEmptyString(value, "wisdomCrystal", errors, "front.wisdomCrystal");
}

function validateBack(value: unknown, errors: string[]): void {
  if (!isRecord(value)) {
    errors.push("back must be an object");
    return;
  }

  requireNonEmptyString(value, "coreAction", errors, "back.coreAction");
  requireNonEmptyStringArray(value, "whenToUse", errors, "back.whenToUse");
  requireNonEmptyStringArray(value, "realLifeExamples", errors, "back.realLifeExamples");
  requireNonEmptyString(value, "commonTraps", errors, "back.commonTraps");
}

function requireNonEmptyString(
  record: Record<string, unknown>,
  key: string,
  errors: string[],
  path = key
): void {
  if (typeof record[key] !== "string" || (record[key] as string).trim().length === 0) {
    errors.push(`${path} must be a non-empty string`);
  }
}

function requireNonEmptyStringArray(
  record: Record<string, unknown>,
  key: string,
  errors: string[],
  path = key
): void {
  const value = record[key];
  if (!Array.isArray(value) || value.length === 0) {
    errors.push(`${path} must include at least one entry`);
    return;
  }

  if (!value.every((item) => typeof item === "string" && item.trim().length > 0)) {
    errors.push(`${path} must contain only non-empty strings`);
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: StarGuardian/Assets/Scripts/Core/ToolCard.cs
// 智慧结晶卡数据模型 + 校验 —— 从 assets/scripts/core/ToolCard.ts 迁移, 规则不变.
// validateToolCard 的输入是"未知形状的 JSON"(TS unknown) → C# 用 Newtonsoft JToken 承接
// (Unity 官方包 com.unity.nuget.newtonsoft-json / dotnet 测试侧 NuGet, 两边同一库).

using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;

namespace StarGuardian.Core
{
    public sealed class ToolCardFront
    {
        public string ToolName { get; init; } = "";
        public string Scene { get; init; } = "";
        public string WisdomCrystal { get; init; } = "";
    }

    public sealed class ToolCardBack
    {
        public string CoreAction { get; init; } = "";
        public IReadOnlyList<string> WhenToUse { get; init; } = Array.Empty<string>();
        public IReadOnlyList<string> RealLifeExamples { get; init; } = Array.Empty<string>();
        public string CommonTraps { get; init; } = "";
    }

    /// <summary>无 UnlockedAt 的草稿(config 里的形态) —— TS 的 ToolCardDraft</summary>
    public sealed class ToolCardDraft
    {
        public string PuzzleId { get; init; } = "";
        public int Stage { get; init; }
        public ToolCardFront Front { get; init; } = new();
        public ToolCardBack Back { get; init; } = new();
    }

    public sealed class ToolCard
    {
        public string PuzzleId { get; init; } = "";
        public int Stage { get; init; }
        public ToolCardFront Front { get; init; } = new();
        public ToolCardBack Back { get; init; } = new();
        public long UnlockedAt { get; init; }
    }

    public sealed class ToolCardValidationResult
    {
        public bool Ok { get; }
        public IReadOnlyList<string> Errors { get; }

        private ToolCardValidationResult(bool ok, IReadOnlyList<string> errors)
        {
            Ok = ok;
            Errors = errors;
        }

        public static ToolCardValidationResult Success() =>
            new(true, Array.Empty<string>());

        public static ToolCardValidationResult Failure(IReadOnlyList<string> errors) =>
            new(false, errors);
    }

    public static class ToolCardFactory
    {
        /// <summary>从草稿造卡(拷贝数组, 不共享引用) —— TS createToolCard</summary>
        public static ToolCard Create(ToolCardDraft draft, long unlockedAt)
        {
            return new ToolCard
            {
                PuzzleId = draft.PuzzleId,
                Stage = draft.Stage,
                Front = new ToolCardFront
                {
                    ToolName = draft.Front.ToolName,
                    Scene = draft.Front.Scene,
                    WisdomCrystal = draft.Front.WisdomCrystal
                },
                Back = new ToolCardBack
                {
                    CoreAction = draft.Back.CoreAction,
                    WhenToUse = draft.Back.WhenToUse.ToList(),
                    RealLifeExamples = draft.Back.RealLifeExamples.ToList(),
                    CommonTraps = draft.Back.CommonTraps
                },
                UnlockedAt = unlockedAt
            };
        }

        /// <summary>校验未知 JSON 是否为合法 ToolCard —— TS validateToolCard, 错误文案逐字保持</summary>
        public static ToolCardValidationResult Validate(JToken? value)
        {
            var errors = new List<string>();

            if (value is not JObject obj)
            {
                return ToolCardValidationResult.Failure(new[] { "tool card must be an object" });
            }

            RequireNonEmptyString(obj, "puzzleId", errors);
            var stage = obj["stage"];
            if (stage is not { Type: JTokenType.Integer } || (long)stage! < 1 || (long)stage! > 5)
            {
                errors.Add("stage must be an integer from 1 to 5");
            }

            ValidateFront(obj["front"], errors);
            ValidateBack(obj["back"], errors);

            var unlockedAt = obj["unlockedAt"];
            var isFiniteNumber =
                unlockedAt is { Type: JTokenType.Integer } ||
                (unlockedAt is { Type: JTokenType.Float } && double.IsFinite((double)unlockedAt!));
            if (!isFiniteNumber)
            {
                errors.Add("unlockedAt must be a finite number");
            }

            return errors.Count > 0
                ? ToolCardValidationResult.Failure(errors)
                : ToolCardValidationResult.Success();
        }

        /// <summary>校验已构造的卡(常用捷径: 序列化后走同一套 JSON 校验, 单一真源)</summary>
        public static ToolCardValidationResult Validate(ToolCard card) =>
            Validate(JToken.FromObject(new
            {
                puzzleId = card.PuzzleId,
                stage = card.Stage,
                front = new
                {
                    toolName = card.Front.ToolName,
                    scene = card.Front.Scene,
                    wisdomCrystal = card.Front.WisdomCrystal
                },
                back = new
                {
                    coreAction = card.Back.CoreAction,
                    whenToUse = card.Back.WhenToUse,
                    realLifeExamples = card.Back.RealLifeExamples,
                    commonTraps = card.Back.CommonTraps
                },
                unlockedAt = card.UnlockedAt
            }));

        private static void ValidateFront(JToken? value, List<string> errors)
        {
            if (value is not JObject obj)
            {
                errors.Add("front must be an object");
                return;
            }

            RequireNonEmptyString(obj, "toolName", errors, "front.toolName");
            RequireNonEmptyString(obj, "scene", errors, "front.scene");
            RequireNonEmptyString(obj, "wisdomCrystal", errors, "front.wisdomCrystal");
        }

        private static void ValidateBack(JToken? value, List<string> errors)
        {
            if (value is not JObject obj)
            {
                errors.Add("back must be an object");
                return;
            }

            RequireNonEmptyString(obj, "coreAction", errors, "back.coreAction");
            RequireNonEmptyStringArray(obj, "whenToUse", errors, "back.whenToUse");
            RequireNonEmptyStringArray(obj, "realLifeExamples", errors, "back.realLifeExamples");
            RequireNonEmptyString(obj, "commonTraps", errors, "back.commonTraps");
        }

        private static void RequireNonEmptyString(JObject obj, string key, List<string> errors, string? path = null)
        {
            var token = obj[key];
            if (token is not { Type: JTokenType.String } || string.IsNullOrWhiteSpace((string?)token))
            {
                errors.Add($"{path ?? key} must be a non-empty string");
            }
        }

        private static void RequireNonEmptyStringArray(JObject obj, string key, List<string> errors, string path)
        {
            if (obj[key] is not JArray array || array.Count == 0)
            {
                errors.Add($"{path} must include at least one entry");
                return;
            }

            var allNonEmptyStrings = array.All(item =>
                item.Type == JTokenType.String && !string.IsNullOrWhiteSpace((string?)item));
            if (!allNonEmptyStrings)
            {
                errors.Add($"{path} must contain only non-empty strings");
            }
        }
    }
}
